### PR TITLE
multiple links added

### LINKED ISSUE #3187: multiple link fields needed under Documentation for API UI
we need to have multiple links for documentation, this is particular need for VRK / SOAP

### PR DESCRIPTION
**Pull request by krashna-deligence**
Closes #3187

## Changes
Describe your changes here as a bulleted list:
- Array added
- Iterate array in HTML
- Create multiple input tag with max limit 8

## Developer checklist
This checklist is to be completed by the PR developer:

- [ ] Alternative solutions were compared/discussed before writing code
    - [ ] trade-offs with this solution are considered acceptable
- [ ] Code in this PR adheres to the [project styleguide](CONTRIBUTING.md)
- [ ] This pull request does not decrease project test coverage
- [ ] If the code changes existing database collection(s), migration has been written
- [ ] If UI texts are added or changed, all texts are internationalized

## Reviewer checklist
Reviewed by: @55

This list is to be completed by the pull request reviewer:
- [ ] Code works as described/expected
- [ ] Code seems to be error free
    - [ ] no browser console errors visible
    - [ ] no server console errors visible
    - [ ] passes CI build
- [ ] Code is written in a way that promotes maintainability
    - [ ] easy to understand
    - [ ] well organized
    - [ ] follows project coding standards and conventions
    - [ ] well documented

### CODE REVIEW

**Comment by 55** [comment]
@krashna-deligence all builds must pass before merging.

**Comment by 55** [comment]
Since there can be more than one link, change text to plural. (link -> links)

d swagger petstore

File: apinf_packages/api_docs/client/view/documentation.html [hunk 1]
```diff
@@ -57,12 +57,14 @@ <h2 id="documentation-link-title" class="panel-title">
...
-                  <i class="fa fa-external-link"></i>&nbsp;
-                  <a href="{{ apiDoc.otherUrl }}" target="_blank">
-                    {{ apiDoc.otherUrl }}
-                  </a>
-                </div>
+                {{# each otherUrl }}
+                  <div id="documentation-link" class="documentation-link">
+                    <i class="fa fa-external-link"></i>&nbsp;
+                    <a href="{{ this }}" target="_blank">
+                      {{ this }}
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/view/documentation.html]: I think it is a bad decision and not human readable, not explicit what did you do here. 

**Remake it using `each-in` construction**

File: apinf_packages/api_docs/client/view/documentation.html [hunk 2]
```diff
@@ -57,12 +57,14 @@ <h2 id="documentation-link-title" class="panel-title">
             <div class="panel-body">
               <!-- Display link to remote documentation-->
               {{# if apiDoc.otherUrl }}
-                <div id="documentation-link" class="documentation-link">
-                  <i class="fa fa-external-link"></i>&nbsp;
-                  <a href="{{ apiDoc.otherUrl }}" target="_blank">
-                    {{ apiDoc.otherUrl }}
-                  </a>
-                </div>
+                {{# each otherUrl }}
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/view/documentation.html]: And why did you add this helper if `otherUrl` is available from the template like `apiDoc.otherUrl`?

**Remove redundant helper**

File: apinf_packages/api_docs/client/view/documentation.html [hunk 3]
```diff
@@ -57,7 +57,7 @@ <h2 id="documentation-link-title" class="panel-title">
             <div class="panel-body">
               <!-- Display link to remote documentation-->
               {{# if apiDoc.otherUrl }}
-                {{# each otherUrl }}
+                {{# each apiDoc.otherUrl }}
                   <div id="documentation-link" class="documentation-link">
                     <i class="fa fa-external-link"></i>&nbsp;
                     <a href="{{ this }}" target="_blank">
```
Comment (krashna-deligence) [inline apinf_packages/api_docs/client/view/documentation.html]: Hi @marla-singer i used `this` to access element from `apiDocs` (plain array).

Comment (marla-singer) [reply apinf_packages/api_docs/client/view/documentation.html]: Why did you not use `apiDoc.otherUrl` directly?

Comment (krashna-deligence) [reply apinf_packages/api_docs/client/view/documentation.html]: Because `apiDocs.otherUrl` gives empty result inside `{{# each apiDoc.otherUrl }}`.

Comment (marla-singer) [reply apinf_packages/api_docs/client/view/documentation.html]: When you have to work with it as with Array. In JS there are forEach & for loops. In the spacebars there is 
```html
{{# each iteration in Array }}
```
http://blazejs.org/guide/reusable-components.html#Prefer-lt-￼16-gt

**Remake it using the correct operator each-in**

File: apinf_packages/api_docs/collection/schema.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ ApiDocs.schema = new SimpleSchema({
     regEx: SimpleSchema.RegEx.Url,
   },
   otherUrl: {
-    type: String,
+    type: [String],
```
Comment (marla-singer) [inline apinf_packages/api_docs/collection/schema.js]: Becaus of you changes the schema, you need to write a new migration step for old documents

**Steps for reproduce**
1. `git checkout develop`
2. Create an API and add a link to Documentation (not swagger)
3. Check out to your branch
4. Go to previously API profile

**Expected result**
- The link is displayed

**Actual result**
- The error on the browser console
```
Error: {{#each}} currently only accepts arrays, cursors or falsey values.
    at badSequenceError (observe-sequence.js?hash=8fe58036c6ba00c458f54c360a21fd0e41fb7ee0:183)
```

File: apinf_packages/api_docs/client/manage/manage.html [hunk 5]
```diff
@@ -71,7 +71,7 @@ <h1 class="modal-title">
               {{> showHelp 'documentation_link' }}
               {{_ "manageApiDocumentationModal_DocumentationLink_Title" }}
             </legend>
-            {{> afQuickField name='otherUrl' }}
+            {{> afQuickField name='otherUrl' maxCount=8 }}
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.html]: Remove `maxCount: 8` value from this. Add this option to related field in collection schema

**Comment by matleppa** [comment]
I think there is a problem with the UI used to add and remove external links.

The look-and-feel and way, how it is used differs from previous UIs used on similar purpose. 
In addition to that there is some weird things happening, which may confuse the user.

After some additions and removals there are empty slots. And for some reason the empty slots are located in different places in UI form and in link list.

image

The logic differs from previous UIs for similar purpose. The '+' and '-' signs are not used elsewhere on our site. Here pressing '+' or '-' opens of closes an input field and the operation needs to be done separately by pressing the [Save] button.

An example of our current logic can be found e.g. addition of Organization managers. (Organizations > Settings).

image

There managers are added one by one using only one input field. 
- Pressing [Add] button adds the new manager at once and simultaneously updates the list of managers.
- Pressing red button removes the manager from list (of course there is security in order to avoid accidental removal) and from DB. The manager list is updated, no empty slots.

### Conclusion
My opinion is, that in our UI the similar functionalities should be implemented as similar as possible.
@preriasusi What do you say?

**Comment by marla-singer** [comment]
@krashna-deligence @matleppa @55 If we go with new UI then have to close this PR and open another one

**Comment by Nazarah** [comment]
1+ to @matleppa 's suggestion.
I have tested use of this controller in branding page for selecting social media for platform footer.
The controller is buggy and doesn't render the UI accordingly when the contents are modified.
a related bug was reported [here](https://github.com/apinf/platform/issues/3213)

**Comment by saransh-dev** [comment]
So, Can we create new PR by taking base PR of this one ? Please advise

**Comment by marla-singer** [comment]
> So, Can we create new PR by taking base PR of this one ? Please advise

You can continue in this one if it will help

**Comment by saransh-dev** [comment]
>You can continue in this one if it will help
Hi @marla-singer, then @krashna-deligence will start work on new UI in same PR

File: apinf_packages/api_docs/client/manage/manage.js [hunk 6]
```diff
@@ -149,4 +166,43 @@ Template.manageApiDocumentationModal.events({
       templateInstance.removeDocumentationFile(fileId);
     }
   },
+  'click #addLinks': function (event, templateInstance) {
+    // Get Value from textbox
+    const link = $('#linksField').val().trim();
+    // Regex for https protocol
+    const regex = '^http(s?)\:\/\/[0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*'+
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.js]: Use regex from SimpleSchema ```SimpleSchema.RegEx.Url```

**Comment by matleppa** [comment]
How is this supposed to work?

Opened modal Document > Manage
Added a external documentation link into input field.

mult_input_link


After pressing [Add] button.
The added link appeared into list on modal.

mult_inputted_link


But link is not visible on Documentation page.

mult_input_link_where


After re-opening the modal the the external documentation link is not visible in list on modal either.

mult_input_link_where_2

**Comment by krashna-deligence** [comment]
Hi @matleppa , external links will not insert in database while click on Add button, it just store in session. After click on save button it create collection and store in database.

Do i need to store external document links in database by click on Add button?

Please advise.

**Comment by Nazarah** [comment]
@krashna-deligence 
please add the following changes in the UI

image

1. There is very little padding between the bottom most link and Save button.
Add same amount of padding that exists between Top most link and  abd "Links to external documentation" field

2. The maximum limit of external links are supposed to be 8. 
However I could 8 more than 8 links.
Please confirm this and fix (if necessary)

3. Change the tooltip message for ? icon as the following:
"Link to the Wiki, Github page or other web resource where the documentation of your API is located. You can add upto 8 external documentation links."
add the second line if the above case is valid.

4. (in case max number of links is 8)
When user adds the 8th link, make the external documentation field disabled.
You can also disable the Add button.
In both case, add the following tooltip on mouse hover: "Maximum 8 external documentation links are allowed. Please delete from existing ones if you wish to add a new one."

5. use the following UI style to show the added links in the API documentation dialog
image
This is from API profile -> Settings tab
ul class = "List-group"
li class 0 "list-group-item"

**Comment by Nazarah** [comment]
In addition, currently on clicking Save button after adding en external link, the added links don't appear on the External Link section on Documentation tab. 
Please fix this.

**Comment by marla-singer** [comment]
Take it on review, working in process

**Comment by marla-singer** [comment]
Works fine.

1. Place the button "Add" to the right side and make the input longer

File: apinf_packages/core/migrations/server/14-fix-api-docs.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+// Meteor contributed packages imports
+import { Migrations } from 'meteor/percolate:migrations';
+
+// Collection imports
+import apiDocs from '/apinf_packages/api_docs/collection';
+
+Migrations.add({
+  version: 14,
+  name: 'Fix api documents migration: change type File to Url',
```
Comment (marla-singer) [inline apinf_packages/core/migrations/server/14-fix-api-docs.js]: Change the migrate name, it doesn't match with the real logic

File: apinf_packages/core/migrations/server/14-fix-api-docs.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  up () {
+    // If documentation is uploaded as remote file but type is file
+    apiDocs.find({ otherUrl: { $exists: true } }).forEach((doc) => {
```
Comment (marla-singer) [inline apinf_packages/core/migrations/server/14-fix-api-docs.js]: Update comment, it doesn't match with the logic

File: apinf_packages/core/lib/i18n/en.i18n.json [hunk 9]
```diff
@@ -427,19 +427,21 @@
   "manageApiDocumentationModal_CreateDocumentation_CancelButton": "Cancel",
   "manageApiDocumentationModal_CreateDocumentation_SaveButton": "Save",
   "manageApiDocumentationModal_CreateDocumentation_Title": "Create documentation file",
+  "manageApiDocumentationModal_CreateLinks_AddButton": "Add",
   "manageApiDocumentationModal_DeletedFile_ConfirmationMessage": "Are you sure you want to remove the documentation file?",
   "manageApiDocumentationModal_DeletedFile_Message": "Documentation file successfully deleted!",
-  "manageApiDocumentationModal_DocumentationLink_Title": "Link to external documentation resource",
+  "manageApiDocumentationModal_DocumentationLink_Title": "Links to external documentation resource",
   "manageApiDocumentationModal_FileType_Message": "Only YAML (YML) and JSON file formats are supported.",
   "manageApiDocumentationModal_LinkField_Updated_Message": "Documentation link successfully updated!",
```
Comment (marla-singer) [inline apinf_packages/core/lib/i18n/en.i18n.json]: link -> links

File: apinf_packages/api_docs/client/manage/manage.html [hunk 10]
```diff
@@ -71,7 +71,42 @@ <h1 class="modal-title">
...
+                   <fieldset class="form-group">
+                     <label for="links">
+                      {{_ "manageApiDocumentationModal_DocumentationLink_Title" }}
+                     </label>
+                     <div class="row">
+                       <div class="col-sm-6">
+                        <input id="linksField" name="links" class="form-control" data-toggle="tooltip" title="">
+                       </div>
+                       <div class="col-sm-6">
+                         <button type="button" class="btn btn-primary" id="addLinks">
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.html]: addLinks -> add-link

1. A user adds only one link via input then it has to be singular
2. The attributes values have format using hyphen instead of camel case

File: apinf_packages/api_docs/client/manage/manage.html [hunk 11]
```diff
@@ -71,7 +71,42 @@ <h1 class="modal-title">
...
-            {{> afQuickField name='otherUrl' }}
+               {{# if otherUrlsCount }} 
+                 <div >
+                   <fieldset class="form-group">
+                     <label for="links">
+                      {{_ "manageApiDocumentationModal_DocumentationLink_Title" }}
+                     </label>
+                     <div class="row">
+                       <div class="col-sm-6">
+                        <input id="linksField" name="links" class="form-control" data-toggle="tooltip" title="">
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.html]: linksField -> link-value
1. A user adds only one link via input then it has to be singular
2. The attributes values have format using hyphen instead of camel case
3. Remove data-toggle attribute if it doesn't need or provide a title value if the tooltip needs

File: apinf_packages/api_docs/client/manage/manage.html [hunk 12]
```diff
@@ -71,7 +71,42 @@ <h1 class="modal-title">
...
+                      </div>
+                    </div>
+                     <small class="text-danger invisible" id="errorMessage">
+                      {{_ "invalidUrlMessage" }}
+                     </small>
+                   </fieldset>
+                 </div>
+               {{/ if }}
+             {{# each otherUrls }}
+               <div id="otherUrl-link" class="otherUrl-link">
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.html]: Remake loop using `each url in otherUrls` construction

File: apinf_packages/api_docs/client/manage/manage.css [hunk 13]
```diff
@@ -6,3 +6,16 @@ https://joinup.ec.europa.eu/community/eupl/og_page/european-union-public-licence
 .documentation-file {
   margin-bottom: 2em;
 }
+
+.cursor {
+	cursor: pointer;
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.css]: Add indentations for all new styles

File: apinf_packages/api_docs/client/manage/autoform.js [hunk 14]
```diff
@@ -18,26 +18,27 @@ AutoForm.hooks({
...
         // Get ID of uploaded file
         const fileId = Session.get('fileId');
-
         // If selected option is File and File is uploaded
         if (doc.type === 'file' && fileId) {
           // Save ID
           doc.fileId = fileId;
         }
+        if (Session.get('links')) {
+          doc.otherUrl = Session.get('links');
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/autoform.js]: Store Session variable to separated variable and use it

File: apinf_packages/api_docs/client/manage/manage.html [hunk 15]
```diff
@@ -71,7 +71,42 @@ <h1 class="modal-title">
...
+                 <ul class="list-group">
+                   <li class="list-group-item">
+                    <a href="{{ this }}" id="oherUrlLinks{{ @index }}">
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.html]: 1. Add attribute "target=_blank" for `<a>` tag
2. otherUrlLinks@index -> other-url-@index
Because this item displays value only one link

File: apinf_packages/api_docs/client/manage/manage.js [hunk 16]
```diff
@@ -44,6 +45,17 @@ Template.manageApiDocumentationModal.onDestroyed(() => {
   Session.set('fileUploading', undefined);
 });
 
+Template.manageApiDocumentationModal.onRendered(() => {
+  // Fetch other Url
+  const apiDocs = ApiDocs.findOne();
+  if (apiDocs && apiDocs.otherUrl) {
+    // Set Other Url in Session variable
+    Session.set('links', apiDocs.otherUrl);
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.js]: Using `_.get` function from `lodash` llibrary

```js
const apiDocs = ApiDocs.findOne()

const links = _.get(apiDocs, 'otherUrl', []);

Session.set('links', links);
```

File: apinf_packages/api_docs/client/manage/manage.js [hunk 17]
```diff
@@ -100,6 +112,29 @@ Template.manageApiDocumentationModal.helpers({
     // Return spinner status
     return Session.get('fileUploading');
   },
+  otherUrls () {
+    // Return Session
+    if (Session.get('links') && Session.get('links').otherUrl) {
+      return Session.get('links').otherUrl;
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.js]: After changes Session variable 'links' will contains the array of links or empty array. 

Change the logic of this helper

File: apinf_packages/api_docs/client/manage/manage.js [hunk 18]
```diff
@@ -100,6 +112,29 @@ Template.manageApiDocumentationModal.helpers({
...
+    }
+    return Session.get('links');
+  },
+  otherUrlsCount () {
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.js]: Incorrect using of helper

1. You can manipulate with attributes using `autorun` function. Session variable is reactive and `autorun` function will run every time after changing Session variable. Put all logic about manipulation with attributes to autorun function
2. Rename helper to `userCanAddUrl` or `availableAddUrls` or the same that will explain the correct meaning

File: apinf_packages/api_docs/client/manage/manage.js [hunk 19]
```diff
@@ -149,4 +184,39 @@ Template.manageApiDocumentationModal.events({
...
+      }
+      // clear the text box
+      $('#linksField').val('');
+    } else {
+      // Hide error message
+      $('#errorMessage').removeClass('invisible');
+    }
+  },
+
+  'click .cursor': function (event) {
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.js]: Rename class value. It doesn't explain what kind of action it is.

File: apinf_packages/api_docs/client/manage/autoform.js [hunk 20]
```diff
@@ -18,26 +18,27 @@ AutoForm.hooks({
       insert (doc) {
         // Get ID of uploaded file
         const fileId = Session.get('fileId');
-
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/autoform.js]: Return this empty line

File: apinf_packages/api_docs/client/manage/autoform.js [hunk 21]
```diff
@@ -18,26 +18,27 @@ AutoForm.hooks({
...
+          doc.otherUrl = Session.get('links');
+        }
 
         // Return data
         return doc;
       },
       update (doc) {
         // Get ID of uploaded file
         const fileId = Session.get('fileId');
-
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/autoform.js]: Return this empty line

File: apinf_packages/api_docs/client/manage/manage.html [hunk 22]
```diff
@@ -69,9 +69,43 @@ <h1 class="modal-title">
...
+                     <label for="links">
+                      {{_ "manageApiDocumentationModal_DocumentationLinks_Title" }}
+                     </label>
+                     <div class="row">
+                       <div class="col-sm-10">
+                        <input id="link-value" name="links" class="form-control" title="">
+                       </div>
+                       <div class="col-sm-2">
+                         <button type="button" class="btn btn-primary" id="add-link" title={{ userCanAddUrl }}>
+                            <span class="glyphicon glyphicon-plus"></span>
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/manage/manage.html]: Use FontAwesome icons instead of plyphicon.

**Remake**

File: apinf_packages/api_docs/client/swagger_ui/swagger_ui.js [hunk 23]
```diff
@@ -34,7 +34,6 @@ Template.swaggerUi.onCreated(() => {
 
       // Get "schemes" property or set an empty array on default
       const schemes = _.get(result, 'schemes', []);
-
```
Comment (marla-singer) [inline apinf_packages/api_docs/client/swagger_ui/swagger_ui.js]: Return this line
